apps/prairielearn/src/pages/instructorStudentDetail/components/StudentAuditEventsTable.tsx

```diff
@@ -0,0 +1,71 @@
+import { FriendlyDate } from '../../../components/FriendlyDate.js';
+import { type StaffAuditEvent } from '../../../lib/client/safe-db-types.js';
+import { type SupportedActionsForTable } from '../../../models/audit-event.types.js';
+
+interface StudentAuditEventsTableProps {
+  events: StaffAuditEvent[];
+}
+
+function renderEnrollmentEventText(event: StaffAuditEvent): string {
+  const { action_detail } = event;
+
+  if (!action_detail) {
+    return 'Enrollment event';
+  }
+
+  const detailMap: Record<SupportedActionsForTable<'enrollments'>, string> = {
+    implicit_joined: 'Enrolled in course',
+    explicit_joined: 'Enrolled in course',
+    invited: 'Invited to course',
+    invitation_accepted: 'Accepted invitation',
+    invitation_rejected: 'Rejected invitation',
+    blocked: 'Blocked from course',
+    unblocked: 'Unblocked from course, now enrolled',
+    invitation_deleted: 'Invitation cancelled', // You can never actually see this state.
+  };
+
+  const detail = detailMap[action_detail as SupportedActionsForTable<'enrollments'>];
+  if (!detail) {
+    throw new Error(`Unknown action detail: ${action_detail}`);
+  }
+  return detail;
+}
+
+export function StudentAuditEventsTable({ events }: StudentAuditEventsTableProps) {
+  if (events.length === 0) {
+    return (
+      <div class="card-body">
+        <div class="text-muted">No enrollment events found.</div>
+        <div class="card-footer text-muted small">
+          Missing events? Enrollment events were not logged before October 2025.
+        </div>
+      </div>
+    );
+  }
```

Comment:
This has some strange styling:

<img width="583" height="175" alt="Image" src="https://github.com/user-attachments/assets/c7ce8ad6-3e29-4189-88c8-d5dbd154889b" />

You can't nest a `card-footer` in a `card-body`.